### PR TITLE
Refactor scrypt hashing and source app session mgmt pt 1/3

### LINKED ISSUE #5455: Improve logic for generate_unique_codename
It's always better and more protected to use while with conditional instead of using `while True`. I came across this while working on #2043 and realized that `generate_unique_codename` actually uses an infinite loop. 

https://github.com/freedomofpress/securedrop/blob/96debdc98be53bafac89b5ddaea761a0f71894fa/securedrop/source_app/utils.py#L40-L46

I took a different approach and used the while with conditional pattern in `display_id`.  If for some reason, the `if` condition gets changed in code and it's not visible in the review that we have an infinite loop depending on that if conditional, we might end up with an infinite loop in the code.

Proposed  solution:
Use the similar logic pattern as in #5381

### PR DESCRIPTION
**Pull request by nabla-c0d3**
## Status

Ready.

## Description of Changes

This PR started with the goal of refactoring **CryptoUtil.hash_codename()** to move it out of CryptoUtil (for #5599).

However, I ran into some problems as I was trying to do this, and saw some opportunities for larger improvements. The downside is that the changes are bigger, so I split them into 3 PRs so they're a bit easier to review. While I know big changes are annoying to review, merge and ultimately deal with, I think these ones are important and worth it.

Overall I felt that a crucial concept was "missing"/invisible in the code base: a user who is currently logged into the source application. 

So the main change of these PRs is the introduction of a **SourceUser** class, which ties it all together: how a user's session is managed (when they create an account, login, etc.), how the user's crypto works (filesystem_id, gpg_secret, etc.), and the user's corresponding "Source" record in the DB. 

My overall goal is to make all this logic self-documenting so it is clear to the reader how these users (and more generally the source app) work. Currently, this logic is spread throughout the code base and also repeated a lot, which makes it hard to understand and hard to change. The main example of this is how hash_codename() is used in a lot of code locations, but in each of them it is hard to understand what is getting hashed, and why.

While I split the changes into 3 PRs, each PR brings something useful independently of the other ones:

* 1/3: This PR: Implement SourceUser and create_source_user() to consolidate source creation (More details below)
* 2/3: https://github.com/freedomofpress/securedrop/pull/5694 Remove CryptoUtil.hash_codename(), scrypt Python dependency, and refactor session management in the source app.
* 3/3: https://github.com/freedomofpress/securedrop/pull/5695 Remove usage of flask.g in the source app and fix tests.

If everything gets merged, the following issues will be completed or will see progress: #5455, #5599, #5613. I also think the source app's code base will be significantly easier to approach, understand, and maintain (at least that's the main goal).

**About this PR which is part 1/3:**

* Implement SourceUser and create_source_user() to consolidate source creation. In the current code base there is some duplicate code (in the tests, the apps, etc.) for creating a new source. This PR replaces all of them with a call to create_source_user() so there's one consistent way of creating a source.
* Fixes https://github.com/freedomofpress/securedrop/issues/5455.

### CODE REVIEW

File: securedrop/crypto_util.py
Comment (nabla-c0d3) [inline securedrop/crypto_util.py:175]: This makes the logic self-documenting and the docstring is no longer needed.

File: securedrop/models.py
Comment (nabla-c0d3) [inline securedrop/models.py:466]: Needed for https://github.com/freedomofpress/securedrop/issues/5613

File: securedrop/passphrases.py
Comment (nabla-c0d3) [inline securedrop/passphrases.py:15]: Only updated a variable's name in this file.

File: securedrop/source_app/main.py
Comment (nabla-c0d3) [inline securedrop/source_app/main.py:80]: This consolidates the app's behavior when something goes wrong during account creation.

File: securedrop/source_app/utils.py
Comment (nabla-c0d3) [inline securedrop/source_app/utils.py]: genkeypair() now takes a source user.

File: securedrop/source_user.py
Comment (nabla-c0d3) [inline securedrop/source_user.py:79]: The scrypt hashing logic becomes private and an "implementation detail" of how the source app and users work.

File: securedrop/tests/test_source_user.py
Comment (nabla-c0d3) [inline securedrop/tests/test_source_user.py:93]: The test vectors in this test were generated using the previous / non-cryptography implementation of scrypt, to ensure that the new implementation returns the same hashes when given the same input.

**Comment by emkll** [comment]
Hi @nabla-c0d3 , thank you for your continued contributions to the project. We are currently preparing a release this week and next, but will get around to providing comments/review to this PR after that. Given the size and the areas of the code this PR is affecting, it may take us a while to review this series of proposed changes.

**Comment by nabla-c0d3** [comment]
> Given the size and the areas of the code this PR is affecting, it may take us a while to review this series of proposed changes.

@emkll No problem and makes sense!

**Comment by eloquence** [comment]
Hi @nabla-c0d3, thanks a lot for taking on this complex refactoring work. I just want to manage expectations on review -- we're currently very focused on finishing Ubuntu 20.04 support, and these PRs touch a lot of key areas of the code, so it may still take us several weeks to complete a thorough review of these changes. That said, if you have any questions in the meantime, please don't hesitate to ping.

**Comment by nabla-c0d3** [comment]
No rush on my side

**Comment by eloquence** [comment]
Thanks for rebasing @nabla-c0d3 -- we're just coming out of the 1.8.0 release cycle and should be able to take a closer look in the next few days, once tests are passing. (Let us know if we can help with that.)

**Comment by nabla-c0d3** [comment]
@eloquence Sounds good. 

The test that is failing doesn't seem to be related to my changes. The latest run on the develop branch failed in a similar fashion (different test but same error "Directory not empty": https://app.circleci.com/pipelines/github/freedomofpress/securedrop/2119/workflows/c6371546-78d2-4050-8513-b33960083a7d/jobs/52269).

I will try rebasing again on the next commit that gets merged into develop, and hopefully the tests will pass,

**Comment by nabla-c0d3** [comment]
Only one test failed, but I don't think it's related to my changes. AFAIK I haven't touched apparmor or anything related to it.

File: securedrop/loaddata.py
Comment (redshiftzero) [inline securedrop/loaddata.py]: `source.pending`

File: securedrop/source_app/main.py
Comment (redshiftzero) [reply securedrop/source_app/main.py:80]: behavior change from the perspective of the user when `SourcePassphraseCollisionError` is raised since they will now get a flashed message, this is fine since the `codename` is no longer in the session when the exception is raised

File: securedrop/tests/test_source.py
Comment (redshiftzero) [inline securedrop/tests/test_source.py:170]: note to other maintainers: this PR does preserve the duplicate tab behavior implemented in the test steps in #5075

File: securedrop/tests/test_source_user.py
Comment (redshiftzero) [inline securedrop/tests/test_source_user.py]: nit: slight preference for an all caps constant for clarity at the top of the file or at the top of the function, e.g. `TEST_SALT_GPG_SECRET` for the values for `salt_for_gpg_secret` and similar for `salt_for_filesystem_id`

File: securedrop/tests/test_source_user.py
Comment (redshiftzero) [reply securedrop/tests/test_source_user.py:93]: sick

File: securedrop/loaddata.py
Comment (redshiftzero) [inline securedrop/loaddata.py]: should be taking a `SourceUser` instance now?

**Review by redshiftzero — COMMENTED** [review]
hey thanks for this! I definitely agree the `SourceUser` object makes the logic clearer. dropped a few notes inline, nits are optional

File: securedrop/loaddata.py
Comment (nabla-c0d3) [reply securedrop/loaddata.py]: Good catch

File: securedrop/loaddata.py
Comment (nabla-c0d3) [reply securedrop/loaddata.py]: Correct - good catch

Comment (nabla-c0d3) [reply securedrop/loaddata.py]: Not detected by mypy due to https://github.com/freedomofpress/securedrop/issues/5599

**Comment by eloquence** [comment]
@nabla-c0d3 We'll aim to take another pass through this PR in the next two weeks (@kushaldas has offered to be point) - we'll probably not be able to merge it yet, as we'll ideally merge all 3 PRs together. Thanks again for all your work on this.

**Comment by nabla-c0d3** [comment]
@eloquence Sounds good. There is one test failing when running `make testinfra` (ie. not the app's code) but I haven't been able get `testinfra` to run at all on a fresh VM.

**Comment by eloquence** [comment]
@nabla-c0d3 If you have cycles, would you mind rebasing your branch against `develop` once more? Major code removal just landed which will also ensure that tests pass for future changes. Thanks, and sorry this is taking a while - big changes :)

**Comment by nabla-c0d3** [comment]
> 
> 
> @nabla-c0d3 If you have cycles, would you mind rebasing your branch against `develop` once more? Major code removal just landed which will also ensure that tests pass for future changes. Thanks, and sorry this is taking a while - big changes :)

No problem - I just did the rebase and it went smoothly. There are two tests failing in testinfra now, but as mentioned before I have not been able to setup a VM that can run `make testinfra`.

**Comment by nabla-c0d3** [comment]
I have rebased this PR, but the same `test_aa_no_denies_in_syslog()` is still failing.

I haven't been able to figure it out because the CircleCI log does not show the full output (`Full output truncated (4869 lines hidden), use '-vv' to show`). I also haven't been able to get `make testinfra` to work in an Ubuntu VM - all the tests in testinfra just fail right away.
Being able to see the full output of the test would already help a lot, but I am not sure how to get to that.

**Comment by zenmonkeykstop** [comment]
Hi @nabla-c0d3 - looks like a file `securedrop/source_user.py` was added - for staging and prod environments, new files need to be explicitly included in the apache2 apparmor profile at `install_files/ansible-base/roles/build-securedrop-app-code-deb-pkg/files/usr.sbin.apache2`. Not sure what's going on with `make testinfra`, will look into it, but hopefully CI should get a little further now.

(Also rebased again to pull in an unrelated CI fix)

**Comment by nabla-c0d3** [comment]
> for staging and prod environments, new files need to be explicitly included in the apache2 apparmor profile

Thanks for looking into this. Looks like CI is now passing! 🎉

File: securedrop/tests/test_manage.py
Comment (kushaldas) [inline securedrop/tests/test_manage.py:227]: `codename` was not used, so this is fine.

**Review by kushaldas — DISMISSED** [review]
There is one comment from @redshiftzero https://github.com/freedomofpress/securedrop/pull/5692/files#r601688777 which can be addressed in this while doing the conflict fix. After that we can merge this.

To test 
- [x] I created a staging instance from `develop` and created users.
- [x] created sources on that
- [x] build debian package from this branch and redeployed
- [x] old users can login, new users can be created and login properly

The code changes look good to me. @nabla-c0d3 after you update the branch, I will do one quick round of check and merge. For now moving to the next PR. Thank you once again for doing this.

**Comment by nabla-c0d3** [comment]
@kushaldas Thanks for looking into this and testing it.

> There is one comment from @redshiftzero https://github.com/freedomofpress/securedrop/pull/5692/files#r601688777 which can be addressed in this while doing the conflict fix. After that we can merge this.

Ok I have made the change and fixed the conflict.

File: securedrop/tests/test_source_user.py
Comment (nabla-c0d3) [inline securedrop/tests/test_source_user.py:14]: Moved these to the top of the file

File: securedrop/source_user.py
Comment (zenmonkeykstop) [inline securedrop/source_user.py]: Though the chance of this being hit is vanishingly small (and logging is disabled by default for the SI), IMO passphrases of sources shouldn't be logged.

**Review by zenmonkeykstop — CHANGES_REQUESTED** [review]
One minor point re: exception error messages

File: securedrop/source_user.py
Comment (nabla-c0d3) [reply securedrop/source_user.py]: Yeah I agree. I removed the passphrase from the exception's message and put the filesystem_id instead.

**Review by zenmonkeykstop — APPROVED** [review]

**Pull request merged**
